aten/src/ATen/native/quantized/cuda/fake_quantize_per_channel_affine.cu

```diff
@@ -0,0 +1,143 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/cuda/CUDAApplyUtils.cuh>
+#include <cmath>
+#include "fake_quantize_core.h"
+
+/* FakeQuantize Op for PerChannelAffine quantization scheme */
+namespace at {
+namespace native {
+
+
+  /* Per channel fake-quantizes the 'inputs' tensor.
+  Args:
+    self: Forward input tensor.
+    scale: scale of per channel affine quantization
+    zero_point: zero_point of per channel affine quantization
+    axis: int specifying the axis to be quantized
+    quant_min: minimum quantized value
+    quant_max: maximum quantized value
+  Returns:
+    Fake quantized tensor (double dtype).
+
+  */
+Tensor fake_quantize_per_channel_affine_cuda(
+    const Tensor& self,
+    const Tensor& scale,
+    const Tensor& zero_point,
+    int64_t axis,
+    int64_t quant_min,
+    int64_t quant_max) {
+
+
+  TORCH_CHECK(self.is_cuda());
+  TORCH_CHECK(self.scalar_type() == ScalarType::Float);
+
+  TORCH_CHECK(scale.size(0) == zero_point.size(0),
+  "scale and zero-point need to have the same dimensions");
+  TORCH_CHECK(scale.size(0) == self.size(axis),
+  "dimensions of scale and zero-point are not consistent with input tensor")
+
+
+  TORCH_CHECK(
+      quant_min <= quant_max,
+      "`quant_min` should be less than or \
+        equal to `quant_max`.");
+
+
+  TORCH_CHECK(at::min(zero_point).item().toLong() >= quant_min &&
+              at::max(zero_point).item().toLong() <= quant_max,
+      "`zero_point` must be between `quant_min` and `quant_max`.");
+
+  TORCH_CHECK(axis >= 0 &&
+              axis <= self.dim(),
+      "`axis` must be between 0 and number of dimensions of input");
+
+  auto Y = at::empty_like(self);
+  for (int i = 0; i < self.size(axis); i++)
+  {
+    auto X_slice = self.slice(axis,i,i+1);
+    auto Y_slice = Y.slice(axis,i,i+1);
+    float sc = scale[i].item().toFloat();
+    int64_t zp = zero_point[i].item().toLong();
+    fake_quantize_slice_cuda(Y_slice, X_slice, sc, zp, quant_min, quant_max);
```

Comment:
you could declare this function with REGISTER_DISPATCH from DispatchStub.h and they you don't need to duplicate the top-level function at all. You'd just do REGISTER_CUDA_DISPATCH (and you'd move CPU implementation to per-architecture folder)

Reply:
Will mark as a TODO